File: Scr_Java/estructuras/Mapa.java
package estructuras;

import modelo.Coordenada;
import modelo.Torre;

public class Mapa {
    private final Coordenada[] ruta;
    private final Coordenada fuenteEnergia;
    private final Torre[] torres;
    private int cantidadTorres;

    public Mapa(Coordenada[] ruta, int capacidadTorres) {
        this.ruta = ruta;
        this.fuenteEnergia = ruta[ruta.length - 1];
        this.torres = new Torre[capacidadTorres];
        this.cantidadTorres = 0;
    }

    // Pre: hay espacio en el arreglo y la casilla no esta ocupada por otra torre.
    public boolean colocarTorre(Torre t) {
        if (cantidadTorres >= torres.length) {
            return false;
        }
        if (buscarTorreEn(t.getPosicion()) != null) {
            return false;
        }
        torres[cantidadTorres] = t;
        cantidadTorres++;
        return true;
    }

    public void quitarTorre(Torre t) {
        int indice = -1;
        for (int i = 0; i < cantidadTorres; i++) {
            if (torres[i] == t) {
                indice = i;
                break;
            }
        }
        if (indice == -1) {
            return;
        }
        // Compacta el arreglo desplazando los elementos posteriores una posicion.
        for (int i = indice; i < cantidadTorres - 1; i++) {
            torres[i] = torres[i + 1];
        }
        torres[cantidadTorres - 1] = null;
        cantidadTorres--;
    }

    // Usado por la opcion "mejorar torre" del menu para ubicar la torre
    // por su posicion (se apoya en Coordenada.equals).
    public Torre buscarTorreEn(Coordenada pos) {
        for (int i = 0; i < cantidadTorres; i++) {
            if (torres[i].getPosicion().equals(pos)) {
                return torres[i];
            }
        }
        return null;
    }

    // Dibuja la grilla: E = entrada de la ruta, F = fuente de energia (final
    // de la ruta), letra inicial del tipo = torre, . = casilla libre.
    public void imprimirMapa() {
        int filas = 0;
        int columnas = 0;
        for (Coordenada c : ruta) {
            filas = Math.max(filas, c.getFila() + 1);
            columnas = Math.max(columnas, c.getColumna() + 1);
        }
        for (int i = 0; i < cantidadTorres; i++) {
            Coordenada p = torres[i].getPosicion();
            filas = Math.max(filas, p.getFila() + 1);
            columnas = Math.max(columnas, p.getColumna() + 1);
        }

        char[][] grilla = new char[filas][columnas];
        for (char[] fila : grilla) {
            java.util.Arrays.fill(fila, '.');
        }
        Coordenada entrada = ruta[0];
        grilla[entrada.getFila()][entrada.getColumna()] = 'E';
        grilla[fuenteEnergia.getFila()][fuenteEnergia.getColumna()] = 'F';
        for (int i = 0; i < cantidadTorres; i++) {
            Coordenada p = torres[i].getPosicion();
            grilla[p.getFila()][p.getColumna()] = torres[i].getTipo().charAt(0);
        }

        for (char[] fila : grilla) {
            StringBuilder sb = new StringBuilder();
            for (char c : fila) {
                sb.append(c).append(' ');
            }
            System.out.println(sb.toString().trim());
        }
    }

    public Coordenada[] getRuta() {
        return ruta;
    }

    public Coordenada getFuenteEnergia() {
        return fuenteEnergia;
    }

    public Torre[] getTorresActivas() {
        return java.util.Arrays.copyOf(torres, cantidadTorres);
    }
}

File: cpp/estructuras/Mapa.cpp
#include "Mapa.h"
#include <iostream>
#include <sstream>
#include <algorithm>

// Guarda la ruta tal cual se la pasan, la fuente de energia es la ultima
// casilla de esa ruta, y el arreglo de torres arranca lleno de nullptr.
Mapa::Mapa(const std::vector<Coordenada>& ruta_, int capacidadTorres_)
    : ruta(ruta_),
      fuenteEnergia(ruta_.back()),
      torres(capacidadTorres_, nullptr),
      capacidadTorres(capacidadTorres_),
      cantidadTorres(0) {}

bool Mapa::colocarTorre(Torre* t) {
    // No hay espacio: ya se llego a la capacidad maxima.
    if (cantidadTorres >= capacidadTorres) {
        return false;
    }
    // Ya hay otra torre en esa misma casilla.
    if (buscarTorreEn(t->getPosicion()) != nullptr) {
        return false;
    }
    // Se agrega al final del "arreglo" activo y se aumenta el contador.
    torres[cantidadTorres] = t;
    cantidadTorres++;
    return true;
}

void Mapa::quitarTorre(Torre* t) {
    // Primero se busca en que indice esta la torre.
    int indice = -1;
    for (int i = 0; i < cantidadTorres; i++) {
        if (torres[i] == t) {
            indice = i;
            break;
        }
    }
    // Si no se encontro, no hay nada que hacer.
    if (indice == -1) {
        return;
    }
    // Se recorre desde esa posicion hacia adelante, moviendo cada torre
    // un lugar hacia atras (esto "tapa" el hueco que deja la que se quito).
    for (int i = indice; i < cantidadTorres - 1; i++) {
        torres[i] = torres[i + 1];
    }
    // La ultima posicion activa queda libre otra vez.
    torres[cantidadTorres - 1] = nullptr;
    cantidadTorres--;
}

Torre* Mapa::buscarTorreEn(const Coordenada& pos) const {
    // Recorre solo la parte activa del arreglo (0 .. cantidadTorres-1) y
    // compara posiciones con equals().
    for (int i = 0; i < cantidadTorres; i++) {
        if (torres[i]->getPosicion().equals(pos)) {
            return torres[i];
        }
    }
    return nullptr;
}

void Mapa::imprimirMapa() const {
    // Primero se calcula cuantas filas y columnas necesita la grilla,
    // revisando tanto la ruta como las torres colocadas (por si alguna
    // esta mas alla del propio camino).
    int filas = 0;
    int columnas = 0;
    for (const Coordenada& c : ruta) {
        filas = std::max(filas, c.getFila() + 1);
        columnas = std::max(columnas, c.getColumna() + 1);
    }
    for (int i = 0; i < cantidadTorres; i++) {
        const Coordenada& p = torres[i]->getPosicion();
        filas = std::max(filas, p.getFila() + 1);
        columnas = std::max(columnas, p.getColumna() + 1);
    }

    // Se arma una grilla de caracteres, todo lleno de '.' (casilla libre)
    // al empezar.
    std::vector<std::vector<char>> grilla(filas, std::vector<char>(columnas, '.'));

    // Se marcan la entrada ('E') y la fuente de energia ('F').
    const Coordenada& entrada = ruta.front();
    grilla[entrada.getFila()][entrada.getColumna()] = 'E';
    grilla[fuenteEnergia.getFila()][fuenteEnergia.getColumna()] = 'F';

    // Se marca cada torre con la primera letra de su tipo ('A' de
    // Arquero, 'C' de Canon).
    for (int i = 0; i < cantidadTorres; i++) {
        const Coordenada& p = torres[i]->getPosicion();
        grilla[p.getFila()][p.getColumna()] = torres[i]->getTipo()[0];
    }

    // Se imprime fila por fila, con un espacio entre cada caracter para
    // que se vea como grilla y no como texto pegado.
    for (const std::vector<char>& fila : grilla) {
        std::ostringstream sb;
        for (char c : fila) {
            sb << c << ' ';
        }
        std::string linea = sb.str();
        // Se quita el espacio extra que queda al final de la linea.
        while (!linea.empty() && linea.back() == ' ') {
            linea.pop_back();
        }
        std::cout << linea << std::endl;
    }
}

const std::vector<Coordenada>& Mapa::getRuta() const { return ruta; }
const Coordenada& Mapa::getFuenteEnergia() const { return fuenteEnergia; }

// Copia solo la parte del arreglo que esta realmente en uso (los
// primeros cantidadTorres elementos), dejando fuera los nullptr sobrantes.
std::vector<Torre*> Mapa::getTorresActivas() const {
    return std::vector<Torre*>(torres.begin(), torres.begin() + cantidadTorres);
}
